### PR TITLE
[hist] Fix TH1::KolmogorovTest with option X when one histogram has zero errors

### PR DESCRIPTION
**Pull request by lmoneta**
Fix the case of using KS test when one of the histogram has zero errors (i.e is a function). Improve KS test adding possibility to specify number of toys by using option "X=number" for example "X=1000"

This PR fixes  the issue #13697

### CODE REVIEW

File: hist/hist/src/TH1.cxx [hunk 1]
```diff
@@ -8230,11 +8247,17 @@ Double_t TH1::KolmogorovTest(const TH1 *h2, Option_t *option) const
...
       // make nEXPT experiments (this should be a parameter)
       prb3 = 0;
+      if (afunc1) Copy(*h1Expt);
+      if (afunc2) h2->Copy(*h2Expt);
       for (Int_t i=0; i < nEXPT; i++) {
-         h1Expt->Reset();
-         h2Expt->Reset();
-         h1Expt->FillRandom(h1_cpy, (Int_t)esum1);
-         h2Expt->FillRandom(h1_cpy, (Int_t)esum2);
+         if (!afunc1) {
```
Comment (guitargeek) [inline hist/hist/src/TH1.cxx:8253]: Can we introduce a comment here, explaining what happens in the loop over over toys and why we don't need to fill the histograms if `afunc1` or `afunc2` respectively? And Maybe it's also good to note that they can't be both `false`, which is checked at the beginning of the function. Otherwise one might wonder why the loop is even implemented also for this case.

File: hist/hist/src/TH1.cxx [hunk 2]
```diff
@@ -8230,11 +8247,17 @@ Double_t TH1::KolmogorovTest(const TH1 *h2, Option_t *option) const
 
       // make nEXPT experiments (this should be a parameter)
       prb3 = 0;
+      if (afunc1) Copy(*h1Expt);
+      if (afunc2) h2->Copy(*h2Expt);
```
Comment (guitargeek) [inline hist/hist/src/TH1.cxx:8251]:
```suggestion
```
Is this even necessary? The `h1Exp` is already created as a clone of `this`, and the `h2Exp` could also be created as a clone of `h2`.

By the way, when you create the clones, I would use a lambda to reduce code duplication and also use `std::unique_ptr`, replacing the block a few lines before with:
```c++
      auto cloneHist = [&](TH1 const *hist) {
         return std::unique_ptr<TH1>{static_cast<TH1 *>((gDirectory ? gDirectory : gROOT)->CloneObject(hist, false))};
      };
      std::unique_ptr<TH1> h1_cpy = cloneHist(this);
      std::unique_ptr<TH1> h1Expt = cloneHist(this);
      std::unique_ptr<TH1> h2Expt = cloneHist(this /*or maybe h2 here?*/);
```

File: hist/hist/src/TH1.cxx [hunk 3]
```diff
@@ -8230,11 +8247,17 @@ Double_t TH1::KolmogorovTest(const TH1 *h2, Option_t *option) const
...
-         h2Expt->Reset();
-         h1Expt->FillRandom(h1_cpy, (Int_t)esum1);
-         h2Expt->FillRandom(h1_cpy, (Int_t)esum2);
+         if (!afunc1) {
+            h1Expt->Reset();
+            h1Expt->FillRandom(h1_cpy, (Int_t)esum1);
+         }
+         if (!afunc2) {
+            h2Expt->Reset();
+            h2Expt->FillRandom(h1_cpy, (Int_t)esum2);
```
Comment (guitargeek) [inline hist/hist/src/TH1.cxx:8259]: Why is `h2Expt` a randomized version of `h1_cpy` here? Should it not be drawn from h2?

**Review by guitargeek — CHANGES_REQUESTED** [review]
Hi Lorenzo, thanks for the PR! I have a few comments about code style, code commenting, redundant copying and a potential problem of randomizing h2.

File: hist/hist/src/TH1.cxx (hunk 2 above)
Comment (lmoneta) [reply hist/hist/src/TH1.cxx:8251]: Thanks for the comment. I will simplify the code using just copy and avoid un-neeeded clones of histograms

File: hist/hist/src/TH1.cxx (hunk 3 above)
Comment (lmoneta) [reply hist/hist/src/TH1.cxx:8259]: No, here we need to find the distribution of the KS test statistics, dsEXPT for the NULL, i.e. when we draw two histograms from the same parent distribution. 
However, you are right there is an error. If afunc1=TRUE we need to draw from h2. So the logic is not correct.

**Review by guitargeek — APPROVED** [review]
LGTM! Thanks a lot for the fix and addressing the review comments!
